polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp

```diff
@@ -730,14 +938,21 @@ static bool hasConflictsInLoop(Operation &op, LoopLikeOpInterface loop,
                  << "can be hoisted: conflicting operation will be hoisted\n");
       continue;
     }
+
+    if (versionCandidate.isVersionCandidate(op, other)) {
```

Comment:
This is strange because looks like the `SYCLAccessorVersionBuilder` represents both the class that create the versioned loop as well as the candidate operations that require versioning in order to be hoisted. Can we have 2 classes, one to represent the builder and one to represent the candidate operation that require versioning ? 

I think we need: 
 - a loop versioning builder which is responsible to version a loop, it should receive the Loop to be versioned and the (aggregated) condition to use to version the loop.
 - a LICM candidate, which should know the condition it needs in order for itself to be hoisted. 

Once we have those 2 classes we need 'something' that collect all candidates and create the aggregated condition to give to the loop version builder.

Reply:
Created https://github.com/intel/llvm/pull/8950 to add a `LICMCandidate` class first.